forge/ee/lib/billing/index.js

```diff
@@ -531,6 +531,57 @@ module.exports.init = async function (app) {
             subscription.trialEndsAt = null
             subscription.trialStatus = app.db.models.Subscription.TRIAL_STATUS.ENDED
             await subscription.save()
+        },
+
+        /**
+         * Resynchronises a subscription with stripe. This ensures are local view
+         * of the subscription status matches stripe's view.
+         */
+        resyncTeamSubscription: async (team, subscription) => {
+            if (subscription.subscription) {
+                try {
+                    const stripeSubscription = await stripe.subscriptions.retrieve(subscription.subscription)
+                    if (stripeSubscription.status !== subscription.status) {
+                        // We have got out of sync. We primarily care about being
+                        // in the canceled state.
+                        // Note: a canceled subscription cannot become uncanceled
+                        if (stripeSubscription.status === 'canceled') {
+                            await app.billing.updateSubscriptionStatus(subscription, stripeSubscription.status, team)
+                            await app.db.controllers.Team.suspendTeam(team)
+                        }
+                    }
+                } catch (err) {
+                    // For now, we'll log this error, but not rethrow it.
+                    app.log.warn(`Error syncing team ${team.hashid} subscription ${subscription.subscription}: ${err.message}`)
+                }
+            }
+        },
+
+        /**
+         * Update a subscription's local status. This is typically going to be done
+         * when we are detect a change from stripe.
+         *
+         * In addition to updating the field, it adds an audit log entry of the change.
+         *
+         * @param {*} subscription
+         * @param {*} newStatus
+         * @param {*} team
+         * @param {*} user
+         */
+        updateSubscriptionStatus: async function (subscription, newStatus, team, user) {
```

Comment:
This feel like it should live in the subscription model or be tied to sequelize, but I guess we need access to the user who performed the action, so not quite as straightforward.

Reply:
I agree, however it requires access to the top level `app` object (for the audit logging functions) which models don't by default, so this felt like the next best place to put it for now.